dev/scripts/src/alluxio.org/build/cmd/docker.go

```diff
@@ -0,0 +1,167 @@
+ */
+
+package cmd
+
+import (
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/palantir/stacktrace"
+	"gopkg.in/yaml.v3"
+
+	"alluxio.org/build/artifact"
+	"alluxio.org/command"
+)
+
+const (
+	defaultDockerYmlFilePath = "src/alluxio.org/build/docker.yml"
+
+	tempAlluxioTarballName        = "alluxio-tmp.tar.gz"
+	tempAlluxioTarballPlaceholder = "${ALLUXIO_TEMP_TARBALL}"
+)
+
+type DockerImage struct {
+	BuildArgs  []string `yaml:"buildArgs,omitempty"`
+	BuildDir   string   `yaml:"buildDir"`
+	Dockerfile string   `yaml:"dockerfile"`
+	Tag        string   `yaml:"tag"`
+	TargetName string   `yaml:"targetName,omitempty"`
+	Dependency string   `yaml:"dependency,omitempty"`
+
+	outputTarball string `yaml:"-"`
+}
+
+type dockerBuildOpts struct {
+	*buildOpts
+
+	dockerImages  map[string]*DockerImage
+	dockerYmlFile string
+	image         string
+	tarballPath   string
+}
+
+func DockerF(args []string) error {
+	cmd := flag.NewFlagSet(Docker, flag.ExitOnError)
+	opts := &dockerBuildOpts{}
+	// docker flags
+	cmd.StringVar(&opts.dockerYmlFile, "dockerYmlFile", defaultDockerYmlFilePath, "Path to docker.yml file")
+	cmd.StringVar(&opts.image, "image", "", "Choose the docker image to build. See available images in docker.yml")
+	cmd.StringVar(&opts.tarballPath, "tarballPath", "", "Set to use existing tarball for building docker image")
+	// parse flags
+	tOpts, err := parseTarballFlags(cmd, args)
+	if err != nil {
+		return stacktrace.Propagate(err, "error parsing build flags")
+	}
+
+	// parse available docker images in docker.yml
+	{
+		wd, err := os.Getwd()
+		if err != nil {
+			return stacktrace.Propagate(err, "error getting current working directory")
+		}
+		dockerYmlPath := filepath.Join(wd, opts.dockerYmlFile)
+		content, err := ioutil.ReadFile(dockerYmlPath)
+		if err != nil {
+			return stacktrace.Propagate(err, "error reading file at %v", dockerYmlPath)
+		}
+		if err := yaml.Unmarshal(content, &opts.dockerImages); err != nil {
+			return stacktrace.Propagate(err, "error unmarshalling docker images from:\n%v", string(content))
+		}
+	}
+
+	image, ok := opts.dockerImages[opts.image]
+	if !ok {
+		return stacktrace.NewError("must provide valid 'image' arg")
+	}
+	alluxioVersion, err := alluxioVersionFromPom()
+	if err != nil {
+		return stacktrace.Propagate(err, "error parsing version string")
+	}
+	opts.buildOpts = tOpts
+	if opts.artifactOutput != "" {
+		artifact, err := artifact.NewArtifact(
+			artifact.DockerArtifact,
+			opts.outputDir,
+			strings.ReplaceAll(opts.dockerImages[opts.image].TargetName, versionPlaceholder, alluxioVersion),
+			alluxioVersion,
+			map[string]string{"image": opts.image})
+		if err != nil {
+			return stacktrace.Propagate(err, "error adding artifact")
+		}
+		return artifact.WriteToFile(opts.artifactOutput)
+	}
+
+	if opts.tarballPath == "" {
+		if err := buildTarball(tOpts); err != nil {
```

Comment:
defer delete tarball if built on demand? or even better if it's directly built into the expected location within dockerWs so it gets deleted in the existing defer